### PR TITLE
[None][feat] Refactor Llava-Next

### PR DESCRIPTION
**Pull request by yechank-nvidia**
This PR refactors the Llava-Next into single executor mode.

## Summary by CodeRabbit

* **New Features**
  * Enhanced support for multimodal (text and image) processing with a new vision model component.
  * Added a new parameter for image processing customization.

* **Refactor**
  * Centralized vision-related logic into a dedicated vision model, simplifying and improving input handling for multimodal tasks.
  * Streamlined preprocessing and feature extraction for images, improving consistency and maintainability.

* **Bug Fixes**
  * Updated expected keywords in multimodal image test to reflect improved model output accuracy.

* **Chores**
  * Updated input processor and model initialization to better support advanced multimodal workflows.

### CODE REVIEW

File: tensorrt_llm/_torch/models/modeling_llava_next.py [hunk 1]
```diff
@@ -44,7 +49,45 @@ def __init__(self,
...
+            text=text_prompt,
+            images=mm_data['image'],
+            do_rescale=not isinstance(mm_data['image'][0], torch.Tensor),
+            return_tensors="pt")
+        # Postprocess
+        fused_input_ids = processed_values['input_ids'][0]
+        if not mm_data:
+            return fused_input_ids.to(torch.int32).tolist(), {}
+        fused_input_ids[fused_input_ids ==
+                        self.image_token_index] = self.vocab_size + 1
```
Comment (amukkara) [inline tensorrt_llm/_torch/models/modeling_llava_next.py:72]: why do we need to remap `image_token_index` to `vocab_size + 1`? is this specific to llava-next?

File: tensorrt_llm/_torch/models/modeling_llava_next.py [hunk 2]
```diff
@@ -78,145 +127,132 @@ def __init__(self,
...
 
         # Use HF multi-modal projector
         self.mm_projector = hf_mm_projector
+        self.image_newline = hf_image_newline
+        self.vision_feature_select_strategy = getattr(
+            model_config.pretrained_config, "vision_feature_select_strategy",
+            "default")
+
+    # Copied from https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava_next/modeling_llava_next.py#L284
+    def pack_image_features(self,
```
Comment (amukkara) [inline tensorrt_llm/_torch/models/modeling_llava_next.py:144]: please add a short comment on the operation or transform this function performs. can be copied from transformers modeling_llava_next.py

**Review by amukkara — APPROVED** [review]
LGTM

File: tensorrt_llm/_torch/models/modeling_llava_next.py (hunk 2 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_llava_next.py:144]: It is instance method so I couldn't import. I will leave a comment.

File: tensorrt_llm/_torch/models/modeling_llava_next.py [hunk 3]
```diff
@@ -278,13 +315,16 @@ def forward(
         multimodal_params = kwargs.get("multimodal_params", [])
         mm_embeds = []
         if len(multimodal_params) > 0:
-            mm_embeds = [
-                multimodal_param.multimodal_data["multimodal_embedding"]
-                for multimodal_param in multimodal_params
-            ]
-
+            if not DISAGG:
+                mm_embeds = self.mm_encoder.forward(multimodal_params)
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_llava_next.py:319]: Yes. Tested with `mixture_text_image` and worked.

File: tensorrt_llm/_torch/models/modeling_llava_next.py [hunk 4]
```diff
@@ -78,145 +127,132 @@ def __init__(self,
...
-        }
+        image_features = torch.split(image_features, image_num_patches, dim=0)
+        image_features, feature_lens = self.pack_image_features(
+            image_features,
+            image_sizes,
+            vision_feature_select_strategy=self.vision_feature_select_strategy,
+            image_newline=self.image_newline,
+        )
+        image_features = torch.cat(image_features, dim=0)
+        return [image_features]
```
[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_llava_next.py:255]: Answered above.

File: tensorrt_llm/_torch/models/modeling_llava_next.py (hunk 2 above)
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_llava_next.py:150]: Added.

**Review by amukkara — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_llava_next.py [hunk 5]
```diff
@@ -78,145 +127,132 @@ def __init__(self,
...
-                           images=image,
-                           do_rescale=not isinstance(images[0], torch.Tensor),
-                           return_tensors="pt",
-                           device=self.device)['pixel_values'][0].to(
-                               self.device) for image in images
+    @torch.inference_mode()
+    def forward(self, multimodal_params: List[MultimodalParams]):
+        pixel_values = [
+            multimodal_param.multimodal_data["image"]["pixel_values"]
+            for multimodal_param in multimodal_params
```
[the remark this answers is not in the document]
Comment (chang-l) [reply tensorrt_llm/_torch/models/modeling_llava_next.py:214]: >the len(multimodal_params) will be 0

Are you sure? I thought you stripped mm_params for generation, [here](https://github.com/NVIDIA/TensorRT-LLM/blob/6a3a9212846ecaba91d7f436adfca9d92f1658c6/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1263), but `mm_params.multimodal_data` is still presented as an empty dict (for each gen req)...  Pls correct me if wrong

[the remark this answers is not in the document]
Comment (yechank-nvidia) [reply tensorrt_llm/_torch/models/modeling_llava_next.py:214]: Yes. `strip_for_generation` removes all the data except mrope_config(which is only used in Qwen2/2.5-VL) and makes the `multimodal_params` empty dictionary. And after that I only append to do the multimodal_params_list if it has content with `has_content`([link](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tensorrt_llm/inputs/multimodal.py#L425), [link](https://github.com/NVIDIA/TensorRT-LLM/blob/6a3a9212846ecaba91d7f436adfca9d92f1658c6/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1267-L1268)). 

Therefore, except Qwen2/2.5-VL models, the len(multimodal_params) will be 0.

**Review by chang-l — APPROVED** [review]
